RFC: Add workspaces to Cargo

alexcrichton opened the pull request:
Improve Cargo's story around multi-crate single-repo project management by
introducing the concept of workspaces. All packages in a workspace will share
`Cargo.lock` and an output directory for artifacts.

Cargo will infer workspaces where possible, but it will also have knobs for
explicitly controlling what crates belong to which workspace.

[Rendered](https://github.com/alexcrichton/rfcs/blob/cargo-workspace/text/0000-cargo-workspace.md)

sfackler commented:
If I have a manually configured workspace, how will cargo figure out that a crate that I'm compiling belongs to a workspace? Will I have had to have built the workspace root first?

alexcrichton commented:
When Cargo [constructs a workspace](https://github.com/alexcrichton/rfcs/blob/cargo-workspace/text/0000-cargo-workspace.md#constructing-a-workspace) it's built upon the condition that all crates in a workspace can transitively reach all others. This means that for any workspace (manually configured or not) any crate can find its way to the root quickly. In that sense it won't matter if you've built the workspace root first or not, Cargo will be able to build any crate in the workspace at any point in time into the workspace root's output folder.

Does that make sense? Not sure if that answers your question...

gkoz commented:
Are multi-repo workspaces too niche to be considered? 

I've been abusing local path overrides for a while now. It would be nice if "local" workspaces could incorporate that. Imagine two repos `foo` and `bar` side by side, `bar/Cargo.toml` containing
```
[dependencies.foo]
git = "some://where/foo.git"
```

I'd be happy if the "local" workspace could transparently change that to `path = "../foo"` and also use a common target directory.

sfackler commented:
Let's say that I have a project structure like this:

```
.
├── .git
│   ├── ...
├── bar-server
│   ├── Cargo.toml
│   └── src
│       └── main.rs
└── foo
    ├── Cargo.toml
    └── src
        └── lib.rs
```

`bar-server` has a path dependency on `foo`, and wants to be the workspace root. If I'm reading the RFC correctly, `bar-server`'s Cargo.toml will need to explicitly specify `workspace-root = true`. If I take a fresh clone of this repo, go into the foo dir, and build it, how will it know that it needs to look in `bar-server` for the lockfile and target directory? Isn't all of the workspace structure metadata in `bar-server`?

alexcrichton commented:
@gkoz 

That's an interesting idea! This doesn't currently consider multiple disjoint repositories beyond submodules. Using `CARGO_TARGET_DIR` as an environment variable should suffice for that use case, and I'm not sure how idiomatic it is in terms of baking support into Cargo. @wycats may have other opinions though.

---

@sfackler 

Ah yes for the workspace to be constructed there you'd need two bits of configuration, one as `workspace-root = true` in `bar-server` and one as `workspace = ["../bar-server"]` in `foo`. Without the configuration in the `foo` crate the `bar-server` crate will have an implicit workspace edge pointing to `foo`, but because there is no edge pointing back it will not be considered part of the workspace. Once `foo` has been configured, however, it'll be able to figure that out.

sfackler commented:
Ok cool, that makes sense.

nikomatsakis commented:
@alexcrichton 

> Ah yes for the workspace to be constructed there you'd need two bits of configuration, one as workspace-root = true in bar-server and one as workspace = ["../bar-server"] in foo. Without the configuration in the foo crate the bar-server crate will have an implicit workspace edge pointing to foo, but because there is no edge pointing back it will not be considered part of the workspace. Once foo has been configured, however, it'll be able to figure that out.

What happens if people don't set this up correctly? 

It seems a bit disappointing for such a common setup to require explicit configuration; I'm not sure what the alternative would be, though, except for having cargo trawl the git repo to look for things that depend on the current directory. (which...might be ok)

Still, I think i might prefer to have a `Workspace.toml` file that just lists both `foo` and `bar-server`. You might then have multiple such files and an explicit `workspace = ../workspace1.toml` declaration (but otherwise cargo just looks to find at most one, and reports an error if there are multiple).

nikomatsakis commented:
I asked:

> What happens if people don't set this up correctly? 

But I think I get it now. If you don't include all the links, things just aren't considered to be in 1 workspace.

tomaka commented:
I don't like the convention that this RFC suggests:

1. I prefer the hierarchy in @sfackler's example. Putting a crate in a subdirectory of a parent crate can be confusing, as you can't know which directories are part of a crate without looking. For example [here](https://github.com/rust-lang/crates.io/tree/master/src) how can you guess that `s3` is a crate while `user` is a module?

2. Whether the user uses source control should be irrelevant for Cargo. The fact that the project contains a `.git` or a `.svn` directory should be out of scope. For example I don't know want my project to not compile on someone's machine because they downloaded an archive of the project instead of cloning it.

netvl commented:
How does the proposed design works in a scenario when there are multiple binaries and some common library? For example, client-server programs like my [wcd](https://github.com/netvl/wcd) can be structured like this. In wcd I have the following structure:
```
wcd/
  common/
    Cargo.toml
  wcd/
    Cargo.toml, contains [dependencies] common = { path = "../common" }
  wcdctl/
    Cargo.toml, contains [dependencies] common = { path = "../common" }
```

I think I would prefer both binaries to be built together (now it is not very convenient to test how client and server interact when something inside `common` changes, because I have to build and run them separately), but I can't see how exactly they map to workspaces. Should there be two workspaces? Or should I choose one binary crate to be a workspace? Or should `common` be a workspace? All these options seems not very useful/convenient/straightforward to me. I think that the alternative with a virtual root crate probably would work well in this scenario, but I may be mistaken.

nikomatsakis commented:
> I don't like the convention that this RFC suggests

Man, I must have been tired when reading the RFC last night. I didn't really grok that it was suggesting that we should structure our repos in a nested fashion. So, e.g., for LALRPOP, where the main lalrpop crate (`lalrpop`) has a number of subsidiary crates that it uses, I would presumably do something like:

```
lalrpop/
    src/...
    lalrop-intern/
        src/...
```

and in this way, lalrpop would implicitly be the "workspace root"? This makes the name 'root' make more sense to me. :)

I also never thought of structuring things in this way, but I can see that if I did so, then this RFC would allow for a lightweight (no annotation) setup.

alexcrichton commented:
@nikomatsakis 

> What happens if people don't set this up correctly? 

Ah yes, as you surmised the end result is that workspaces just continue to be disjoint (as they are today). If manual configuration is incorrect (e.g. a `workspace` key points to a crate not actually in a workspace) then an error is generated. Any implicit configuration (by structuring the project) is just ignored, however, if it's "incorrect".

---

@tomaka 

> I don't like the convention that this RFC suggests:

One of the major motivational factors for this RFC is to have zero configuration in the conventional use case. A scheme such as you and @sfackler are proposing is certainly covered by this RFC (wish some configuration), but it's unclear to me how it could be the default. The requirement is to be able to go from any crate to all of its workspace crates quickly (eg can't walk the whole filesystem).

Do you have something in mind, however, to alleviate this? To reiterate, this RFC doesn't preclude *any* source layout, it's just blessing one as the conventional way to organize crates where features work without extra configuration.

> Whether the user uses source control should be irrelevant for Cargo.

That's correct, but it's also why it's just a heuristic as part of this RFC. Any and all project layouts are supported, it's just a question of which require less configuration. Most projects usually have some sort of source repo at the top which is a good heuristic for probing (but that's it).

---

@netvl 

> How does the proposed design work in a scenario when there are multiple equally important binaries and some common library?

You'd basically choose one of your crates to be the root, and then you'd have to configure a few `workspace` keys. After that all output would be placed in your chosen root.  It's true that a "virtual package" would perhaps make this somewhat easier to configure because you could place a virtual Cargo.toml at the root of the crate and then everything would automatically be connected to that workspace.

Do you think the configuration necessary to select one of the binaries as the root is too much, though? Is that a case in favor of some form of "virtual package"?

---

@nikomatsakis 

> I also never thought of structuring things in this way, but I can see that if I did so, then this RFC would allow for a lightweight (no annotation) setup.

Yeah that sort of layout would require no extra configuration to benefit, but as @tomaka mentions it may not always make sense for projects. There may be another heuristic we could use to find a "root crate" without any extra configuration, but I'm drawing blanks...

tomaka commented:
An alternative could be to separate workspaces from packages, and create another configuration file named for example `Cargo-workspace.toml`. This is what the Visual Studio IDE does, and I think most IDEs do something similar.

If you want to have one crate per directory, you can just put a `Cargo-workspace.toml` at the root.
It would also cover @netvl's use case.

netvl commented:
@alexcrichton 
> Do you think the configuration necessary to select one of the binaries as the root is too much, though? Is that a case in favor of some form of "virtual package"?

Well, I'm not sure whether it is too much or not, but the whole concept, when applied to "equally important" binaries, seems wrong to me. I just dislike the need to choose one of the crates as a root arbitrarily. Maybe if client-server application is not a convincing example, then something like [coreutils](https://github.com/uutils/coreutils) is? It is a collection of lots of different binaries using a single common library, and I doubt that it is possible to select one of them as a workspace root. Something like cargo-edit also qualifies, I believe. So yes, I certainly think that virtual packages or some equivalent solution is needed here.

withoutboats commented:
I'm a little unclear on the implications of the 'workspace inference' rules. I think the standard way to organize a multi-crate project is this:

```
parent_crate/
  src/
  child_crate_1/
    src/
    grandchild_crate/
      src/
  child_crate_2/
    src/
``` 

My understanding is that this structure will be inferred under the proposal and require no annotation. That sounds great.

But will other structures be inferred? Like @tomaka mentioned, I think the practice of putting multiple crates inside of a `src` directory is confusing and difficult to approach, and I'd rather that if people wanted to do something like that, cargo requires them to provide the explicit annotations. It'd be good in my opinion to have configuration, but _also_ have a clear convention.

Ericson2314 commented:
> @nikomatsakis Still, I think i might prefer to have a `Workspace.toml` file that just lists both foo and bar-server.

Yes! Has anyone looked at Haskell's Stack http://docs.haskellstack.org? IMO whenever I develop multiple crates at once, the idea is almost always that while the libraries are easiest to develop together, they are still full fledged independent libraries:
 - The libraries should be released/uploaded to crates.io seperately
 - Semver is sufficient to force a tight coupling if need be.
 - A patch release in one library shouldn't and won't necessitate a release of the other.

As such a view `Workspace.toml` as something more akin to the overrides in a cargo config file than a normal package. crates.io package should always refer to other libraries on crates.io, but for lockstep development one uses the local version instead.

If the crates are released/uploaded to crates.io separately, they should NOT be nested in the filesystem, otherwise we need to prune sub trees when uploading so one library doesn't contain others' source.

Ericson2314 commented:
The main problem I have stack's `stack.yaml` is that it somewhat takes on the role of a lockfile in that the version control solver is often used to generate it. But this RFC does get that right. `Workspace.toml` has "input constraints", further constraining or overriding the constraints in each cargo file, and `Cargo.lock` (maybe `Workspace.lock` to avoid compatibility issues?), as has always been the case, has the final concrete resolved dependencies.

Ericson2314 commented:
I think @netvl concern's basically boil down to "dependencies form a DAG, but nested crates form a tree". IMO that right there is the simplest reason that nesting crates is going to suck---simpler and better than the problem I described 2 comments previous :).

nikomatsakis commented:
@withoutboats 

> Like @tomaka mentioned, I think the practice of putting multiple crates inside of a src directory is confusing and difficult to approach

Are you saying that the directory structure itself is confusing, or that the annotations one would need (under this RFC) to support it are confusing?

----

@alexcrichton 

I definitely think the RFC would be improved by specifying the expected directory layout. It makes the rules make much more sense. 

Do you know if this "nested" structure is in common use today?

tomaka commented:
I don't know for the nested structure, but non-nested structure are not uncommon:

- https://github.com/bjz/gl-rs
- https://github.com/gfx-rs/gfx/tree/master/src (technically nested, but the root crate has no source code at all)
- https://github.com/serde-rs/serde
- https://github.com/tomaka/vulkano

withoutboats commented:
> Are you saying that the directory structure itself is confusing, or that the annotations one would need (under this RFC) to support it are confusing?

The former. I expect everything inside `src` to be a single crate. There may be good reasons to include other crates inside that directory, but I'd rather for cargo's rules to encourage a narrow set of possible directory structure conventions.

alexcrichton commented:
@tomaka 

> An alternative could be to separate workspaces from packages, and create another configuration file named for example Cargo-workspace.toml.

Yeah this is certainly always possible (and is one of the alternatives in the RFC), but one of the key points here is that an idiomatic project structure should require zero extra configuration (including this extra manifest). 

---

@netvl 

Thanks for the examples! I can certainly see the desire to have a manually selected root location for the workspace, but I'm pretty hesitant to so easily add the concept of virtual crates or packages. It seems like a pretty significant feature that may not quite pull its weight if it's just used for workspaces. For now this could always be worked around with a "dummy package" that ends up being the root, which although not quite as elegant should at least serve as a vector in the interim.

@wycats may have more thoughts on the "virtual package" idea, though. I believe he's thought about it more than I have.

---

@withoutboats 

Could you clarify what you're finding hard to understand about how workspaces are constructed? It's described concretely because that's what'll end up being implemented (I tend to prefer that over a somewhat hand-wavy description). I can certainly add some examples though!

---

@nikomatsakis 

I've seen lots of examples with a nested layout and also lots with a nothing-at-the-root layout. I wouldn't necessarily say one is more common than the other. As I [mentioned in my reply](https://github.com/rust-lang/rfcs/pull/1525#issuecomment-192408632) to @tomaka, however, there's a fundamental problem of child crates still somehow need to quickly discover their parent. That tips the scales in favor (to me) much more in towards a something-should-be-at-the-root structure.

I've tried to add some examples though which may help.

gkoz commented:
Perhaps alternatives should include using `.cargo/config` for this? There already is `target-dir` (which seems to treat relative paths differently from `paths` for some reason). It could also specify a common lockfile.

Ericson2314 commented:
An empty `Workspace.toml` can tell Cargo to just including any `Cargo.toml` in the work-space in the current subtree. Cargo, when walking up the filesystem will look for `Cargo.toml`s and `Workspace.toml`s, preferring the latter. That's nearly zero configuration, and IMO vastly preferable to trying to be clever with `.git`.

withoutboats commented:
> Could you clarify what you're finding hard to understand about how workspaces are constructed? It's described concretely because that's what'll end up being implemented (I tend to prefer that over a somewhat hand-wavy description). I can certainly add some examples though!

It's great that a concrete implementation is provided in the RFC, but some examples of the kinds of directory trees these rules infer and the kinds these rules cannot infer would help to understand how it impacts me as a user.

alexcrichton commented:
@gkoz perhaps, yeah, although `.cargo/config` configuration is generally not intended to be checked into a repository, and most workspace configurations are likely to be checked in. It also doesn't necessarily seem to me like too much of a benefit to just separate out a shared lockfile from a shared artifact folder (and seems like it can easily go wrong as well).

@Ericson2314 yes we could likely infer most configuration from a blank file at a root, but I suspect we will always need some form of VCS recognition to handle cases like git submodules at least. A submodule which has its own workspace needs to continue to be separated from the main repo if included.

@withoutboats I added a few examples to the RFC, are they what you were thinking of?

nikomatsakis commented:
@alexcrichton 

> I've seen lots of examples with a nested layout and also lots with a nothing-at-the-root layout. I wouldn't necessarily say one is more common than the other. 

OK.

> As I mentioned in my reply to @tomaka, however, there's a fundamental problem of child crates still somehow need to quickly discover their parent. That tips the scales in favor (to me) much more in towards a something-should-be-at-the-root structure.

I think what bothers me a bit about the current "workspace-root" and "workspace-edge" annotations (or whatever they were called) is that there is this concept -- the workspace -- but it is not being described directly, but rather in a sort of roundabout fashion. That is, it "emerges" by many independent `Cargo.toml` files being setup in the right way (one of them as root, the others with links to the root), instead of being something I can just describe in one place (a `Workspace.toml`, or maybe a `Cargo.toml` with a `[workspace]` entry, or something like that).

So to put this another way: I think the current set of "implicit edges" seems like a fine heuristic for deducing a workspace, but if people are going to get explicit, it seems simpler to let them just list the packages for a workspace in one central file.

I was talking to @wycats on IRC who mentioned the idea of having a `crates/` directory with a sibling VCS dir be an implicit workspace, and a `Cargo.toml` with a `[workspace]` entry as the explicit form. Any of these things seem pretty ok to me. 

> I've tried to add some examples though which may help.

Yes, thanks.

Ericson2314 commented:
@alexcrichton To achieve that, tweak the rules so that subtrees with their own workspace are excluded. I.e, assuming all `Workspace.toml` are empty. E.g. given:
```tree
├── bar
│   └── Cargo.toml
├── foo
│   └── Cargo.toml
├── submodule
│   ├── baz
│   │   └── Cargo.toml
│   └── Workspace.toml
└── Workspace.toml
```
then the outer workspace just has foo and bar.

Moreover though, I'm not sure we wan't that. I'd think the whole point of submodules is that the inner submodule still stands on its own, but the outer git repository is a major consumer of it such that their are developed together. Therefore, I'd *want* the outer workspace to include the inner one. That could just be my view, but if there isn't a consensus on how best to interpret the VCS data, let's err on the side of less special casing.

Finally many package managers (such as nix) rather not keep the git history of the projects they package around---in the case of Nix this is especially important as the `.git` directory is near-impossible to recreate deterministically when we need a stable hash of the source . Granted, I envision Workspaces being used just for development, not for releases. But if somebody were to make a release with them, it would be extremely annoying for pruning the `.git` directory to change the behavior of Cargo.

alexcrichton commented:
Ok, I've updated this RFC with the major amendment being the addition of a "virtual manifest" which is an otherwise empty `Cargo.toml` except for workspace configurations. This should allow projects as described by @nikomatsakis and @tomaka to simply have a `Cargo.toml` in the root of the project with a `members` key pointing at the root crates of each sub-project.

When running `cargo build` or other commands that *don't* have a `-p` argument against a virtual manifest the command will fail. Arguments with `-p` will, however, continue to work as expected.

SimonSapin commented:
A virtual manifest sounds great for the servo/servo repository. Thanks!

netvl commented:
Do I understand correctly that, even if one is not allowed to run the most useful commands against the workspace root like `build` or `test`, when you compile member crates, you still get the benefits of the shared lockfile and the shared dependencies compilation output? If yes, then it's really great!

alexcrichton commented:
@netvl indeed! You can also work against the root `Cargo.toml` if it's virtual so long as you use the `-p` argument (to select what you want to do).

nikomatsakis commented:
@alexcrichton :+1: from me on the new draft.

nikomatsakis commented:
@alexcrichton can a crate participate in multiple workspaces at once?

I feel like it might be simpler if there was a `project.workspace` key which pointed at the root and then *only* the root `Cargo.toml` had a `[workspace]` entry at all. (And there is no need to define a `workspace.root` key.) Does that make sense?

Put another way:

- projects define edges to their dependencies always
- and optionally one "workspace" edge
  - but we can infer this as long as the workspace is in some ancestor directory

In this case, there is no need to infer when a `Cargo.toml` is a workspace root based on the VCS directory. Basically, every `Cargo.toml` is a workspace root unless one of the following holds:

- there is a `Cargo.toml` with a `[workspace]` entry in some ancestor directory, in which case this project is implicitly in the same workspace as the ancestor;
- it defines a `workspace="foo"` key.

One flaw might be that maybe the existence of a `[workspace]` section is not "significant" in TOML?

alexcrichton commented:
@nikomatsakis 

I... think that your scheme would actually be simpler and continue to cover all use cases proposed in this RFC. I got about halfway through writing a comment about the pros/cons where something like what you proposed might require extra configuration, but I don't think it's actually the case?

Concretely, under what you're thinking, I believe the rules would be:

* Implicit edges of a workspace would be:
  * `path` dependencies
  * If `package.workspace` isn't present, any hierarchical `Cargo.toml` with a `[workspace]` key
* Explicit edges of a workspace would be:
  * `[package.workspace]` entries
  * `[workspace.members]` keys

We would still require that for A and B to be in the same workspace there must be an edge from A to B as well as B to A (transitively), and we would also require that a workspace contained exactly one crate with the `[workspace]` key.

The only downside of this approach that I can think of is that it's impossible for any workspace to implicitly exist. That is, with zero configuration, no workspace can exist. That being said, we can likely augment `cargo new` to generate a top-level Cargo.toml with a `[workspace]` key and all nested `Cargo.toml` files generated could omit the `[workspace]` key by default.

Is that what you were thinking?

Ericson2314 commented:
> can a crate participate in multiple workspaces at once?

> We would still require that for A and B to be in the same workspace there must be an edge from A to B as well as B to A (transitively)

I rather not require symmetry to take the symmetric closure. This trivially allows using the same crate in two different workspaces, because back-references via `package.workspace` are no longer necessary.

> and we would also require that a workspace contained exactly one crate with the `[workspace]` key.

Under the asymmetric approach, and "child" crates' workspace member sets gets unioned with parent crate's. Thus, this restriction is not necessary.

> If package.workspace isn't present, any hierarchical `Cargo.toml` with a `[workspace]` key

So I think we can split this apart: under the asymmetric approach, any `Cargo.toml` can serve as the workspace root, and legacy single-Cargofile setups become one-crate workspaces. The question then becomes: based on the current directory, which `Cargo.toml` is chosen as the workspace root? 

In the spirit of orthogonality, I rather have it have it so `package.workspace` doesn't add any crates to any workspaces, but rather overrides the rules so that if the current `Cargo.toml` was chosen, the `package.workspace` delegates a different `Cargo.toml` to be used instead. A warning is given if the delegated `Cargo.toml` does not include the delegating one as a workspace member, and an error is given if the delegations form a cycle.

With that taken care of, path dependencies are the only source of implicit membership.

Ericson2314 commented:
>This trivially allows using the same crate in two different workspaces, because back-references via `package.workspace` are no longer necessary.

To elaborate, just as git has both `-C`, and `--work-tree` + `--git-dir`, I think Cargo should have `-C`, and `--workspace-root`. Then in a crate included in two different workspaces, one can use the latter flag to specify which workspace to use. `package.workspace` would just be used to annoint one of those workspaces as the default, instead of the singleton workspace with the crate in question.

alexcrichton commented:
@Ericson2314 

[As mentioned in the RFC](https://github.com/alexcrichton/rfcs/blob/cargo-workspace/text/0000-cargo-workspace.md#constructing-a-workspace):

> The intention is to ensure that the set of packages in a workspace is the same regardless of which package is selected to start discovering a workspace from.

This is a crucial property which must hold as Cargo needs to always generate the same `Cargo.lock`. It's also highly desirable as it means that you'll always find the same root and won't unnecessarily recompile anything as you cd around a source tree.

nikomatsakis commented:
@alexcrichton 

> Is that what you were thinking?

Roughly, but I think I still intended to include path dependencies as implicit outgoing edges. This would mean it IS possible to have a workspace with no [workspace] key, if you use the nested configuration. (Still, I'm not sure that's an important property. In particular, I don't know that I want to privilege this nested configuration.) It also means that the workspace root would basically be the "outermost cargo.toml" in the event that there is no explicit root given.

Ericson2314 commented:
> This is a crucial property which must hold as Cargo needs to always generate the same `Cargo.lock`.

Note that under my proposal, build artifacts directories and `Cargo.lock`s are still paired 1-1, so there is no soundness issue. The only issue is ergonomics.

> It's also highly desirable as it means that you'll always find the same root and won't unnecessarily recompile anything as you cd around a source tree.

While I that situation is annoying, on others my proposal is more ergonomic. Consider a crate used in two different workspaces, and due to dependencies in those workspaces, it must be configured mutually incompatibly in each. [The mutual incompatibility means there is no way to combine the two workspaces, I presume.]  It would be annoying to constantly edit `package.workspace` based on which workspace one was working on, rather than `cd` to the relevant workspace root.

Because @nikomatsakis's route requires symmetry rather than taking the symmetric closure, my plan can be viewed as purely a generalization. What was required becomes an optional strategy under my plan to recover the exact same ergonomics as before. On the other hand, my plan allows for things that would be disallowed---like the above---which provide better ergonomics in other situations.

alexcrichton commented:
@nikomatsakis 

> This would mean it IS possible to have a workspace with no [workspace] key

For each of the child crates, to find their root they'd have to walk up the filesystem. They need to know when to stop, though, so isn't that the `Cargo.toml` with a `[workspace]`? Or are you thinking that you'd basically choose the "highest" `Cargo.toml` that you find?

@Ericson2314 

Is there a  use case for a crate in two workspaces? I haven't seen anything in the ecosystem where that would be the case at least.

Ericson2314 commented:
@alexcrichton I was recently experimenting with using per-stage `stack.yaml`s for bootstrapping GHCJS, the haskell->javascript compiler: https://github.com/forked-upstream-packages-for-ghcjs/ghcjs/tree/7202a3989cc6ce06106d4905935a1d94964feb12 . This is analogous to using per-stage virtual crates for per-stage workspaces.

Another (hypothetical) usecase would be testing to make sure that changes to a library are not breaking. One workspace would have a bunch of consumers of the library and the last last release of the library from crate.io, the other would have the same consumers and a local copy of the library.

Or, to refine that hypothetical example, maybe the goal not for the library to have no breaking changes, but just to make sure the consumers in question can be made to work with both the old and new releases. So the library and consumers would be developed concurrently.

nikomatsakis commented:
@alexcrichton 

> For each of the child crates, to find their root they'd have to walk up the filesystem. They need to know when to stop, though, so isn't that the Cargo.toml with a [workspace]? Or are you thinking that you'd basically choose the "highest" Cargo.toml that you find?

I was thinking that you would take the highest Cargo.toml you find, if you don't find a workspace (or `.git` directory) first. But honestly I don't care because I don't like the nested format, and I am fine with making a `[workspace]` entry explicitly. 

I feel like the idea of going from one crate to many "interconnected" crates is already a step where you have to learn new things about cargo -- e.g., path dependencies. And it seems fine that at this point you also learn to add a `[workspace]`. If we structure our docs correctly, this seems ok to me.

jansegre commented:
@alexcrichton 
I'd like to bring up the issue @tomaka mentioned earlier about using VCS in the heuristics, I'm still unsure if I understood it.

If I understand correctly the presence of a `.git` directory will affect how an implicit workspace is discovered for the benefit that a common case doesn't need explicit configuration. But is it possible (or worst, probable), that a common setup relies on `.git` for defining its workspace?

If so that would mean people (or CIs, I think appveyor can be configured to download a tarball instead of cloning) would not be able to download and build a random project they find in crates.io,

Diggsey commented:
The solution that you and @nikomatsakis have converged upon seems like a great idea. I know I've run into problems with multirust where depending on where you invoke cargo from, the build artifacts end up in different target directories.

There are a couple of caveats:
1) I share the concerns that depending on VCS roots will cause problems when you want to build from a tarball or similar. We had a similar discussion before, and I concede that it makes sense to reuse the VCS info for publishing crates, but I'd prefer the build system part of cargo to be as independent as possible from any VCS.
2) Calling them "workspaces" will confuse anyone using RustDT :wink:

alexcrichton commented:
Ok, I've pushed an update taking @nikomatsakis's [comment](https://github.com/rust-lang/rfcs/pull/1525#issuecomment-196902562) into account, essentially what was [previously outlined](https://github.com/rust-lang/rfcs/pull/1525#issuecomment-197138086). This update also namely discusses specifically what happens with lock files and discussed the interaction with [top level overrides](https://github.com/rust-lang/cargo/pull/2385) as [brought up](https://github.com/rust-lang/cargo/pull/2385#issuecomment-202293359).

The resulting configurations feel much more natural to me! Most of the time you just need a `Cargo.toml` with a `[workspace]` key and you're good to go, and even more rarely you'll need to specify `workspace.members`, and *even more rarely* you'll specify `package.workspace`. I think that this should cover almost all project layouts I know of today in terms of the addition of of `[workspace]` somewhere will cause the natural workspace to exist.

To clarify, as well, the VCS root business here is purely to prevent Cargo from wildly walking all over the filesystem. The only heuristic it plays into is to halt the traversal upwards looking for an implicit root.

Ericson2314 commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,362 @@
+
+To alleviate misconfiguration, however, if the `workspace.members`
+configuration key contains a crate which is not a member of the constructed
+workspace, Cargo will emit an error indicating as such.
```
This doesn't fix the underlying problem that path dependencies induce an asymmetric relation. [Honestly, I'd deprecate path dependencies and path overrides for `[replace]` in virtual crate root.]

alexcrichton replied:
I... don't understand this. Path dependencies are not going to be deprecated and the point about asymmetry or symmetry is pretty core to this RFC, it's a part which can't really change without a completely new design.

Ericson2314 replied:
The depended-on crates have no idea what other crates depend in them via a path dependency. So how can we use path dependencies to add crates to the workspace while enforcing symmetry?

alexcrichton replied:
That's my point, though. This RFC requires that A and B being in the same workspace means there's a path from A to B and vice versa, and that point is pretty core to this RFC.

Ericson2314 replied:
I understand the symmetry requirement; I don't understand how path dependencies don't break it. If A has a path dependency on B, and one `cd`s to B and runs `cargo build`, how does Cargo know that A and B are in the same workspace?

Ericson2314 replied:
This is exactly the problem you describe in the "With the implicit relations defined it's possible for a repository to not have a..." paragraph.

Ericson2314 replied:
But I don't understand the following paragraph, the one I quoted, at all 

> To alleviate misconfiguration, however, if the `workspace.members` configuration key contains a crate which is not a member of the constructed workspace, Cargo will emit an error indicating as such.

But `workspace.members` *defines* members of the workspace, how could mistakenly include packages? Maybe you meant the opposite: for example when walking up the FS to find a the workspace root, ensure that it includes the original crate as a workspace member (for symmetry). 

Now that still leaves us with the problem that if cargo is invoked from a "shallow" crate, there could be other deeper crates that would pick the current crate as the workspace route due to the tree walking, but that the shallow crate isn't aware, to find them cargo would have to comb the whole fs subtree.

I brought up path dependencies because they are the worse offender. At least the other search is confined to subtrees or the parent directory chain. But path dependencies could be anywhere on the entire filesystem.

alexcrichton replied:
> how does Cargo know that A and B are in the same workspace?

Can you rephrase your question with respect to [this section](https://github.com/alexcrichton/rfcs/blob/cargo-workspace/text/0000-cargo-workspace.md#constructing-a-workspace)? My answer to your question is this sentence:

> Two crates are then in the same workspace if they both transitively have edges to one another.

The problem later describe is *avoided* because of this restriction.

Ericson2314 replied:
Oh so just chuck all asymmetric edges, be they explicit or implicit or transitive, warning or erring according? Then the "With the implicit relations defined..." paragraph problem won't occur at all.

alexcrichton replied:
> But workspace.members defines members of the workspace, how could mistakenly include packages?

That's what the configuration check is for, If A says B is in a workspace, that's only true if B says A is in its workspace as well. Authors could forget half of that relation, and Cargo will generate an informative error in that situation.

The paragraph you're quoting is indicating that when Cargo encounters that sort of relation via path dependencies it will not run into a bad situation where workspaces are different depending on what crate they're viewed from. The wording seems a little confusing, though, so I'll touch it up.

Ericson2314 replied:
Great, thanks! Totally clear now. :)

Ericson2314 commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,362 @@
+  unfortunately, isn't always efficient to do and it would be unfortunate to
+  have to unconditionally do this.
+
+* Cargo could support "virtual packages" where a `Cargo.toml` is placed at the
```
Is out of date? We are doing this, right?

Ericson2314 commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,362 @@
+  it's needed at this point.
+
+* Implicit members are currently only path dependencies and a "Cargo.toml next
+  to VCS" traveling upwards. Instead all Cargo.toml members found traveling
```
"upmost Cargo.toml between VCS root and current directory, inclusive" I believe is the current rule?

matklad commented:
Hm, looks like this may require some changes to `cargo metadata` subcommand. Should this be covered by the RFC text?

alexcrichton commented:
Yeah it would likely involve changing slightly, but only to expand the number of crates returned. Currently `cargo metadata` closely mirrors the `Resolve` struct in Cargo (which is in turn an in-memory representation of `Cargo.lock`). I suspect the changes would basically just naturally propagate outwards to `cargo metadata`

alexcrichton commented:
:bell: This RFC is now entering its week-long **final comment period** :bell:

Ericson2314 commented:
Hmm, I was just looking at my toy kernel for the first time a while, and I realized the proposed VCS rules would get in the way of development workflow. Here's the situation:

 - At https://github.com/RustOS-Fork-Holding-Ground/ I have a number of libraries/forks. While I change them as I please for my kernel use, I try to keep them usable for others, so I *don't* want to add any workspace stuff to them specific to kernel to their `Cargo.toml`s

 - https://github.com/RustOS-Fork-Holding-Ground/RustOS/blob/master/Cargo.toml is the Cargo.toml for my kernel which, for "release" purposes, references branches in the libraries/forks

 - For developement, I clone  everything side-by-side in a directory, and also put a `.cargo/config` in that directory to tell Cargo to use the local paths instead. I'd like to replace that `.cargo/config` with a virtual workspace root with `[replace]` overrides.

 - But even though the virtual crate root will lie up the filesystem from each local clone, because the `.git` directory in each local clone, Cargo won't traverse upstream looking for crates at all, and therefore won't find my virtual root.

 - I could switch to making the `RustOS` repo contain everything else as submodules. But:
   - Submodules are annoying
   - The use of `rustos` branches already makes clear what revisions are used in each library so the submodule config would just be redundant information.
   - I'd like not to burden everyone that clones my repos from dealing with submodules. Only developers needs a workspace. A single `Cargo.toml` with deps is fine for somebody that just wants to build it.

gkoz commented:
The above comment applies to `gtk-rs` just as well.

> For developement, I clone everything side-by-side in a directory, and also put a `.cargo/config` in that directory to tell Cargo to use the local paths instead. I'd like to replace that `.cargo/config` with a virtual workspace root with `[replace]` overrides.

I can understand unwillingness to make `cargo` search for `Cargo.toml` in parent directories. I could live with controlling this through `.cargo/config` e.g. `cargo` reads `workdir/Cargo.toml` only if `workdir/.cargo/config` says `walk_up_here = true`.

tsion commented:
I think the worry about Cargo "walking all over the filesystem" is unfounded, since it'll just end up stepping through a few more parent directories until it hits `/`. I think most people don't have that many directories between `/` and their development directories that it would cause a human-perceptible delay.

So, maybe we can just scrap the VCS stuff to resolve the problems in the last two comments?

alexcrichton commented:
Yes if it causes problems the rule about not walking over VCS directories can be dropped, it's a very minor part of this RFC.

tsion commented:
Sounds good, then! The main idea of the RFC seems quite useful.

brson commented:
Why have two different ways to put a package into workspace? You can either put `package.workspace` in the sub-crate or `workspace.members` in the root. The latter seems more maintainable to me. Why would you want to do the former?

> "If package.workspace isn't specified, then Cargo will walk upwards on the filesystem until either a Cargo.toml with [workspace] is found or a VCS root is found."

Does this mean that in practice you don't have to mention `[workspace]` at all in your Cargo.toml? Since this heuristic is looking for `[workspace]` does that mean you can *just* define the `[workspace]` section without writing the 'members' key?

> "The rust repo conceptually has two workspaces, the standard library and the compiler, and these would need to be manually configured with workspace.members and package.workspace keys amongst all crates."

If there are two local workspaces connected to a third root crate how do you keep the workspaces seperate, and not have them implicitly become one big workspace? Won't the implicit rules walk up the directory and find the VCS root and consider them linked in one workspace?

nrc commented:
I'm in favour of the idea here, but not in favour of the VCS connection. I think it is unnecessary because you can always use one of the phantom Cargo.toml files. I don't like it because:

* it is too implicit, and furthermore implicit in a way that is probably unexpected to someone just considering Cargo
* it ties people's VCS usage into an opinionated workflow. While I don't mind our tools being opinionated about how you use them, I find it weird that they have opinions about how you use other tools.
* which VCS systems do we use and which do we ignore? What happens when new ones appear in the future? Do we know they'll be compatible with any implementation scheme we use now.
* imagine you start a project without VCS, it compiles fine, then you add it all to Git, possibly using some odd configuration and now it doesn't build (or builds differently). This seems extremely surprising to me.

Ericson2314 commented:
@brson You seem to be arguing against the symmetry requirement.

> You can either put `package.workspace` in the sub-crate or `workspace.members` in the root.

I'm pretty sure one always needs `workspace.members` to include the crate, along with either `package.workspace` or the implicit fallback from walking up the filesystem.

> > The rust repo conceptually has two workspaces, the standard library and the compiler

Hmm, I don't think that is possible at all with the symmetry rules?

flavius commented:
I think @netvl's comments should be taken seriously, because it's a common scenario.

matthiasbeyer commented:
This looks like it would really help us over here with [imag](https://github.com/matthiasbeyer/imag) where we have 18 crates in one repository at the time of writing, each crate depending on another.

jnicklas commented:
I agree with @brson's points regarding using VCS information, but I would also like to argue *against* using all `path` dependencies in a workspace automatically. I often use `path` dependencies for debugging and developing purposes, trying out changes in an upstream crate or playing with extracting things into a separate crate. In other words these are path dependencies of the form `../path_to_other_crate`, as such they are *not* part of the same workspace. They will not automatically establish a bi-directional link, and to have one-directional links could potentially be quite confusing.

Instead I'd like to propose the following:

A member crate:

- determines its workspace by traversing upwards until it finds the top-most Cargo.toml file.

- may specify its workspace via `package.workspace`

- may opt out of being in a workspace by setting `package.workspace = false`

- checks that it is a member in the workspace, otherwise any `cargo` command run against it will fail with an appropriate error message.

A workspace root crate:

- finds its member crates by normalizing the paths of all path dependencies and checking if they are subdirectories of the workspace root's directory. Those crate paths are taken as member crates.

- may specify its member crates via `workspace.members`

- walks through all member crates and checks that they point to the correct workspace, if they do not, any command run against either the crate root or any of its members will fail with an appropriate error message.

The key differences from this RFC are:

1. A virtual crate is mandatory in the case where there is no otherwise clear crate root, removing the VCS lookup.

2. Incorrectly setting up the workspace will cause a hard error, instead of simply ignoring the workspace setup.

Because of the way the inference works, there should only be two cases where a user will encounter an error due to workspace configuration:

1. They have set up a custom workspace configuration

2. They are trying to run a cargo command in a crate which is in a subdirectory of another crate but *not* among that crate's members/path dependencies. It should be fairly easy to craft an understandable error message for this case e.g.:

> cargo has inferred that this crate is part of some_crate's workspace at /path/to/crate (because it is in one of its subdirectories), but cannot find it among some_crate's member crates. Either add a dependency on this crate to some_crate or set package.workspace = false in this crate's Cargo.toml.

Ericson2314 commented:
@jnicklas I think your usecase would be better served by a `[replace]` section in the workspace root.

Note also that the RFC is written, incorrectly setup work spaces aren't discarded wholesale, just the asymmetric links. And warnings can still be used---it's not error or silence.

alexcrichton commented:
Ok, the tools team discussed this RFC during its triage meeting the other day, and the conclusion was somewhat mixed. We did not decide to merge just yet, but were somewhat hesitant to do so. There was some pushback about this being an "instantly stable" feature in Cargo because Cargo does not have unstable features like the compiler, and there was also discussion about perhaps blocking this until that existed or somehow ensuring it doesn't reach the stable channel just yet. The question of unstable features in Cargo, however, is pretty weighty, so I'd prefer to not tackle it at this time if possible.

I fear that this RFC was perhaps not well written which led to confusion in how it was interpreted. I've rewritten and reorganized the RFC in a format which I hope is much more clear and much simpler than before. Note, however, that there is no semantic change from the previous draft, just hopefully more understandable!

cc @rust-lang/tools 

---

@brson 

> Why have two different ways to put a package into workspace?

We discussed this in the tools meeting, but I think that the current draft is more clear on this as well. There's only one way to put a package in a workspace, `workspace.members`, and that's it. A *valid workspace* must have members point back to the root, but pointing to a root doesn't make you a member.

> Does this mean that in practice you don't have to mention [workspace] at all in your Cargo.toml?

To clarify, no workspace will exist unless `[workspace]` is mentioned somewhere.

> If there are two local workspaces connected to a third root crate how do you keep the workspaces seperate, and not have them implicitly become one big workspace?

As [pointed out by](https://github.com/rust-lang/rfcs/pull/1525#issuecomment-212146940) @Ericson2314 it's not possible for a crate to be a member of more than one workspace (it can only point to one root)

---

@nrc 

As discussed in the tools triage meeting, the "we don't walk past VCS roots" part has been removed. There is no other VCS related portion of this RFC (to clarify), which I believe addresses your concerns!

---

@jnicklas 

> but I would also like to argue against using all path dependencies in a workspace automatically

The RFC has been changed slightly to account for this (which I think is simpler as well) to say that `path` dependencies are only used if `workspace.members` isn't specified.

> Instead I'd like to propose the following:

Hm I think that the current draft may be a bit more understandable, does it help clear things up? I think it covers many of your points, but I'm not 100% certain.

jnicklas commented:
@alexcrichton the new revision of this RFC is awesome! Really clarifies things and the added explicitness will make things much easier to understand, imo. Definitely addresses my concerns at least!

matthiasbeyer commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+of crates pointing back to the root may not. If, however, this restriction were
+not in place then the set of crates in a workspace may differ depending on
+which crate it was viewed from. For example if workspace root A includes B then
+it will think B is in A's workspace. If, however, B does ont point back to A,
```
Typo: `ont` should be `not`.

(Please tell me if I'm not allowed to point out typos here)

alexcrichton replied:
Thanks!

dcuddeback commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+
+The root of a workspace, indicated by the presence of `[workspace]`, is
+responsible for defining the entire workspace (listing all members).
+This example here means that two extra crates will members of the workspace
```
Typo: "will members" is probably supposed to say "will be members"

dcuddeback commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+These keys are mutually exclusive when applied in `Cargo.toml`. A crate may
+*either* specify `package.workspace` or specify `[workspace]`. That is, a
+crate cannot both be a root in a workspace (contain `[workspace]`) and also be
+member of another workspace (contain `package.workspace`).
```
"be a member" reads more easily to me than "be member"

dcuddeback commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+which crate it was viewed from. For example if workspace root A includes B then
+it will think B is in A's workspace. If, however, B does ont point back to A,
+then B would not think that A was in its workspace. This would in turn cause the
+set of crates in each workspace to be different, futher causing `Cargo.lock` to
```
Typo: "futher" is missing an "r" ("further")

dcuddeback commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+builds no matter where they're executed in the workspace.
+
+To alleviate misconfiguration Cargo will emit an error if the two properties
+above hold for any crate attempting to be part of a workspace. For example, if
```
Is this supposed to say that Cargo will emit an error if the two properties *do not* hold?

nodakai replied:
I think it says no crates other than "the" root should satisfy the conditions in a workspace.

alexcrichton replied:
@dcuddeback oops, indeed!

dcuddeback commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+
+* Test all crates within a workspace (run all unit tests, doc tests, etc)
+* Build all binaries for a set of crates within a workspace
+* Publish all crates in a workspace if necessary to crates.io
```
Related to this use case, another possible extension is to share a version across all the crates in a workspace. Some projects act as a collection of crates that are published at the same time (e.g., `foo` and `foo-sys`) with the same version. It'd be a nice convenience to not have to update the `version` property in every `Cargo.toml`.

dcuddeback replied:
Actually, this could apply to `authors`, `homepage`, `repository`, `license`, and `keywords` properties, too, as those are likely to be the same for crates in a single-repo workspace. `version` just changes more often than the other properties. It would be nice if member crates could inherit these properties from the workspace root.

alexcrichton replied:
Nice idea! I'll add this in here as well.

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+the output of a workspace to be configured regardless of where crates are
+located, Cargo will now allow for "virtual manifest" files. These manifests will
+currently **only** contains the `[workspace]` table and will notably be lacking
+a `[project]` or `[package]` top level key.
```
What was `[project]` ?

alexcrichton replied:
An alias for `[package]` with planned different semantics that never got off the ground.

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+Cargo will for the time being disallow many commands against a virtual manifest,
+for example `cargo build` will be rejected. Arguments that take a package,
+however, such as `cargo test -p foo` will be allowed. Workspaces can eventually
+get extended with `--all` flags so in a workspace root you could execute
```
What are we achieving by requiring `--all` for "virtual" projects?

alexcrichton replied:
I don't think I understand the question here. The point of this is to say that `cargo build` doesn't work, not that `--all` is being added.

nodakai replied:
In other words, why should `cargo build` not work for a "virtual workspace" ? My point is, you are suggesting to have two types of `Cargo.toml`; one accepts `cargo build` and one rejects `cargo build`. This can be a source of confusion.

alexcrichton replied:
Mostly erring on the side of being conservative, we can always alter the meaning of `cargo build` on a virtual crate to be an alias for `cargo build --all` later.

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+### Workspaces in practice
+
+Many Rust projects today already have `Cargo.toml` at the root of a repository,
+and with the small addition of `[workspace]` in the root a workspace will be
```
If I didn't misread the section "Validating a workspace," the workspace is not valid unless we add `[package] workspace = ...` to children crates.

nodakai commented on text/0000-cargo-workspace.md:
````diff
@@ -0,0 +1,345 @@
+  ```
+
+  The two workspaces here can be configured by placing the following in the
+  manifests:
````
Is it correct to understand that the two workspaces just happen to be in the same "tree" (undefined word) and shouldn't share the common `.lock` file? If that is the case, this example seems to me to serve more for confusion rather than explanation and I'd rather not include it in the RFC.

alexcrichton replied:
Yes this is intended to showcase two workspaces as part of the same development tree, which is what the compiler will have, for example.

nodakai replied:
My point is such a situation can arise *too* often, for example, when I check out two unrelated (but workspace-aware) Cargo packages from GitHub under the common directory, say, `~/dev/`. I'd naturally take their independence for granted.

alexcrichton replied:
I'm not sure I understand what change you'd like as a result of these comments then? The intention is that of course two independent trees can have workspaces that don't mess with one another, and the point of this example is simply to show that it can happen in one repo.

nodakai commented on text/0000-cargo-workspace.md:
````diff
@@ -0,0 +1,345 @@
+
+  ```toml
+  # ws2/Cargo.toml
+  [workspace]
````
Shouldn't `members` be here?

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+* A workspace can contain multiple local crates.
+* Each workspace will have a root.
+* Whenever any crate in the workspace is compiled, output will be placed in the
+  `target` directory next to the root.
```
Should we worry about name collision?

alexcrichton replied:
Not any more so than we already have today.

nodakai commented on text/0000-cargo-workspace.md:
````diff
@@ -0,0 +1,345 @@
+  ```
+
+Projects like the compiler will likely need exhaustively explicit configuration.
+The `rust` repo conceptually has two workspaces, the standard library and the
````
I understood a "workspace" as a unit for sharing the common `lock` file. Could you elaborate on why the compiler and the stdlib should not share the common `lock` file?

alexcrichton replied:
It's somewhat unrelated to this RFC, unfortunately, bit the gist of it is that we want crates.io deps to be part of the compiler but they do not explicitly depend on the standard library, so they need to be built in two phases.

nodakai replied:
OK, then I'd suggest you to omit the sentence. You are simply saying "a complex project is likely to require a complex hand-written configuration"

Ericson2314 replied:
There's another solution for that problem ;)

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,345 @@
+Cargo will grow the concept of a **workspace** for managing repositories of
+multiple crates. Workspaces will then have the properties:
+
+* A workspace can contain multiple local crates.
```
Could "multiple" perhaps better be replaced with "zero or more" (if a workspace with zero local crates is not forbidden)?

nodakai commented on text/0000-cargo-workspace.md:
````diff
@@ -0,0 +1,348 @@
+  ```toml
+  # ws1/Cargo.toml
+  [workspace]
+  members = ["crate1", "crate2"]
````
Maybe "Note that `members` aren't necessary ..." ?

alexcrichton replied:
It is required here as the newly created manifest does not otherwise depend on the crates.

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,348 @@
+
+1. A workspace has only one root crate (that with `[workspace]` in
+   `Cargo.toml`).
+2. All workspace crates defined in `workspace.members` point back to the
```
I'd prefer to have "explicitly or implicitly" for `workspace.members` and `package.workspace` for the sake of clarification.

nodakai commented on text/0000-cargo-workspace.md:
````diff
@@ -0,0 +1,348 @@
+```
+
+The root of a workspace, indicated by the presence of `[workspace]`, is
+responsible for defining the entire workspace (listing all members).
````
> listing all members

I'd prefer to have "implicitly or explicitly" here

nodakai commented on text/0000-cargo-workspace.md:
```diff
@@ -0,0 +1,348 @@
+be implicitly defined in some situations.
+
+The `package.workspace` can be omitted if it would only contain `../` (or some
+repetition of it). That is, if the root of a workspace is hierarchically the
```
Consider this case:

```
ws0
 +- Cargo.toml  // [workspace] members = ["ws1/util"]
 +- src
 +- ws1
     +- Cargo.toml // [workspace] members = []
     +- src
     +- util
         +- Cargo.toml // package.workspace is omitted
         +- src
```

So, although `util/../..` does point back to `ws0` and the workspace `ws0` contains a single root, it is invalid due to the "that is" sentence; is my understanding correct?  I mean, the "that is" sentence is actually a stronger restriction than what it tries to rephrase.

alexcrichton replied:
If `ws0/ws1/Cargo.toml` depends on `ws0/ws1/util/Cargo.toml`, then yes this is an invalid workspace.

nodakai replied:
Actually I found "hierarchically the first" confusing; did it mean the closest to the root directory of the system (`/`) or the furthest from it?

I interpreted it as "the furthest" (because it would be the "first" to be found by a repeated application of `../`); then the root of the workspace `ws0` is *not* "the first" `Cargo.toml` with `[workspace]` regardless of `ws0/ws1`'s (lack of) dependency on `ws0/ws1/util`.

My question could be rephrased as: can an unrelated `Cargo.toml` (`ws0/ws1/Cargo.toml` in this case) break the validity of a workspace (`ws0`) that depends on implicit `package.workspace` ?

alexcrichton replied:
Yes

alexcrichton commented:
The tools team discussed this RFC during triage yesterday and the decision was to **merge**. Once this is implemented we're likely to reevaluate right before the next release to see if we want to let it ride the trains to stable, and we may, for one release, temporarily prevent that particular released version of Cargo from having workspaces. This'll all be based on our experience with workspaces up to that point, however!

Thanks again for the discussion everyone!
